src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -0,0 +1,369 @@
+            if (TryReadInternal(tokenSource, out ReadResult readResult))
+            {
+                return readResult;
+            }
+
+            if (_isStreamCompleted)
+            {
+                return new ReadResult(buffer: default, isCanceled: false, isCompleted: true);
+            }
+
+            var reg = new CancellationTokenRegistration();
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.Register(state => ((StreamPipeReader)state).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                var isCanceled = false;
+                try
+                {
+                    AllocateReadTail();
+
+                    Memory<byte> buffer = _readTail.AvailableMemory.Slice(_readTail.End);
+
+                    int length = await InnerStream.ReadAsync(buffer, tokenSource.Token);
+
+                    Debug.Assert(length + _readTail.End <= _readTail.AvailableMemory.Length);
+
+                    _readTail.End += length;
+                    _bufferedBytes += length;
+
+                    if (length == 0)
+                    {
+                        _isStreamCompleted = true;
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearCancellationToken();
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    isCanceled = true;
+                }
+
+                return new ReadResult(GetCurrentReadOnlySequence(), isCanceled, _isStreamCompleted);
+            }
+        }
+
+        private void ClearCancellationToken()
+        {
+            lock (_lock)
+            {
+                _internalTokenSource = null;
+            }
+        }
+
+        private void ThrowIfCompleted()
+        {
+            if (_isReaderCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            ThrowIfCompleted();
+
+            return TryReadInternal(InternalTokenSource, out result);
+        }
+
+        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        {
+            bool isCancellationRequested = source.IsCancellationRequested;
+            if (isCancellationRequested || _bufferedBytes > 0 && (!_examinedEverything || _isStreamCompleted))
+            {
+                // If TryRead/ReadAsync are called and cancellation is requested, we need to make sure memory is allocated for the ReadResult,
+                // otherwise if someone calls advance afterward on the ReadResult, it will throw.
+                if (isCancellationRequested)
+                {
+                    AllocateReadTail();
```

Comment:
`it will throw` why would it throw? Null segment? Can we handle it in `Advance()`?

Reply:
It can be handled in advance, I'm going to change this logic to act more like `Pipe`.